src/Tools/ExternalAccess/VisualDiagnostics/Internal/VisualDiagnosticsServiceFactory.cs

```diff
@@ -0,0 +1,294 @@
+    }
+
+    private class OnInitializedService : ILspService, IOnInitialized, IObserver<HotReloadNotificationType>, IDisposable
+    {
+        private readonly LspServices _lspServices;
+        private readonly LspWorkspaceManager _lspWorkspaceManager;
+        private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
+        private readonly Lazy<IBrokeredDebuggerServices> _brokeredDebuggerServices;
+        private readonly ConditionalWeakTable<Workspace, IVisualDiagnosticsLanguageService> _visualDiagnosticsLanguageServiceTable;
+        private readonly System.Timers.Timer _timer;
+        private readonly SemaphoreSlim _mutex = new SemaphoreSlim(1);
+        private List<ProcessInfo> _debugProcesses;
+        private CancellationToken _cancellationToken;
+        private IDisposable? _adviseHotReloadSessionNotificationService;
+
+        public OnInitializedService(LspServices lspServices, LspWorkspaceManager lspWorkspaceManager, LspWorkspaceRegistrationService lspWorkspaceRegistrationService, Lazy<IBrokeredDebuggerServices> brokeredDebuggerServices)
+        {
+            _lspServices = lspServices;
+            _lspWorkspaceManager = lspWorkspaceManager;
+            _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
+            _brokeredDebuggerServices = brokeredDebuggerServices;
+            _timer = new System.Timers.Timer();
+            _timer.Interval = 1000;
+            _timer.Elapsed += Timer_Elapsed;
+            _visualDiagnosticsLanguageServiceTable = new();
+            _debugProcesses = new List<ProcessInfo>();
+        }
+
+        private void Timer_Elapsed(object sender, System.Timers.ElapsedEventArgs e)
+        {
+            _timer.Stop();
+            _ = OnTimerElapsedAsync();
+        }
+
+        public void Dispose()
+        {
+            _adviseHotReloadSessionNotificationService?.Dispose();
+            (_brokeredDebuggerServices as IDisposable)?.Dispose();
+        }
+
+        public Task OnInitializedAsync(ClientCapabilities clientCapabilities, RequestContext context, CancellationToken cancellationToken)
+        {
+            _cancellationToken = cancellationToken;
+            // This is not ideal, OnInitializedAsync has no way to know when the service broker is ready to be queried
+            // We start a timer and wait roughly a second to see if the broker gets initialized.
+            // TODO dabarbe: Service broker is initialized as part of another LSP service, not sure if there's a way await on, or having a task completion source?
+            _timer.Start();
+            return Task.CompletedTask;
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        public void OnError(Exception error)
+        {
+        }
+
+        public void OnNext(HotReloadNotificationType value)
+        {
+            _ = HandleNotificationAsync(value);
```

Comment:
This needs to use `IAsynchronousOperationListener` since the life of the asynchronous operation is not captured by the caller.